Allow CDM user to read selected CloudWatch logs

jsf9k opened the pull request:
## 🗣 Description ##

This pull request:
- Creates an AWS IAM policy that grants the necessary permissions to read selected CloudWatch logs
- Attaches said policy to the AWS IAM role that the CDM user is already allowed to assume
- Adds @mcdonnnj as a codeowner

See also cisagov/ansible-role-cloudwatch-agent#48.

## 💭 Motivation and context ##

- The CDM folks have requested the ability to read selected CloudWatch logs from our SharedServices instances
- I have been adding @mcdonnnj as a codeowner to any repos where he is not already.  He deserves this honor/burden.

## 🧪 Testing ##

- All automated tests pass
- I have applied these changes to the staging version of the COOL and have visually inspected the IAM resources and verified that they appear as expected
- Once this PR is approved I will go ahead and apply it to the production version of the COOL and ask CDM to verify that they can read the desired CloudWatch logs.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [x] Apply these changes to the production version of the COOL and ask CDM to verify that they function as expected.

## ✅ Post-merge checklist ##

- [x] Update both the staging and production `*.tfvars` files with the new variables.

jsf9k committed:
Add code to allow the CDM user to read selected CloudWatch logs

The CDM folks have requested to be able to read selected CloudWatch
logs from our SharedServices instances.

jsf9k committed:
Add as an output the policy that allows CDM user to read selected CW logs

jsf9k committed:
Update README.md via terraform-docs

jsf9k committed:
Add @mcdonnnj as a codeowner

mcdonnnj commented on cloudwatch_for_cdm.tf:
```diff
@@ -0,0 +1,52 @@
+      # staging is not connected to CDM; this is because CDM cannot
+      # handle the fact that COOL production and staging use the same
+      # IPv4 IP space.
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa0.cool.cyber.dhs.gov:log-stream:messages",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa0.cool.cyber.dhs.gov:log-stream:secure",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa0.cool.cyber.dhs.gov:log-stream:audit",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa1.cool.cyber.dhs.gov:log-stream:messages",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa1.cool.cyber.dhs.gov:log-stream:secure",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa1.cool.cyber.dhs.gov:log-stream:audit",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa2.cool.cyber.dhs.gov:log-stream:messages",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa2.cool.cyber.dhs.gov:log-stream:secure",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/ipa2.cool.cyber.dhs.gov:log-stream:audit",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/vpn:log-stream:messages",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/vpn:log-stream:secure",
+      "arn:aws:logs:${var.aws_region}:${local.sharedservices_account_id}:log-group:/instance-logs/vpn:log-stream:audit",
```
Thoughts on making all of the log groups a `local` and building the list dynamically? Something like:

```terraform
[ for group in local.cloudwatch_log_groups : format("arn:aws:logs:%s:%d:log-group:%s", var.aws_region, local.sharedservices_account_id, group) ]
```

dav3r replied:
I like @mcdonnnj's suggestion - it will make this code a little bit cleaner.  

What I would really like to see is a way for us to remove the hard-coded names of the instances here.  I suppose we could make them Terraform vars or grab them from the outputs of https://github.com/cisagov/cool-sharedservices-freeipa and https://github.com/cisagov/cool-sharedservices-openvpn, but this is complicated by the fact that the log groups for the IPA instances have FQDNs, while those for the OpenVPN instance do not.

jsf9k replied:
Please see commit 1f279bd.  Given that all this CDM code is already specific to our instantiation, I don't have a lot of gumption to figure out a way to produce the instance hostnames from thin air.  That said, I will figure out a way if @dav3r insists.

dav3r replied:
How about a middle ground in the interest of separating configuration from code- can you make `cdm_instances` into a Terraform variable?  If you want to go a step further, I suppose `cdm_log_streams` could also benefit from being a Terraform variable, but I'll leave that one up to you.

jsf9k replied:
Please see commit 691261b.

dav3r replied:
Love it, thanks!

mcdonnnj commented on variables.tf:
```diff
@@ -56,6 +56,18 @@ variable "aws_region" {
+variable "cloudwatch_policy_name" {
+  type        = string
+  description = "The name to assign the IAM policy that allows read access to the specific CloudWatch log streams in which CDM is interested."
+  default     = "CdmCloudWatch"
```
Just a thought since we often throw permission levels in the names of policies I believe.

```suggestion
  default     = "CdmCloudWatchReadOnly"
```

dav3r commented on cloudwatch_for_cdm.tf:
```diff
@@ -0,0 +1,52 @@
+# ------------------------------------------------------------------------------
+# Provision the IAM policy that allows access the CDM CloudWatch data
```
Add a missing word:
```suggestion
# Provision the IAM policy that allows access to the CDM CloudWatch data
```

jsf9k committed:
Add missing word

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Append "ReadOnly" to role name

This makes sense since the role only allows for read-only access.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Generate the list of resources via code instead of by hand

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Change lists of instances and log streams to input variables

This is more flexible, and it also separates code from configuration.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>
Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

dav3r reviewed (approved):
👍 👍 

@jsf9k - Can you also add a pre (or post) merge checklist item to update the staging & prod tfvars files with the new variables?

mcdonnnj reviewed (approved):
:shipit:

jsf9k committed:
Rework logic to only allow CDM access to the duplicate CW logs

In cisagov/ansible-role-cloudwatch-agent#48 we created duplicate CW
log groups for selected logs just for CDM's Splunk ingestion process.
In this commit I rework the logic to only allow access to these
duplicate CW logs.

dav3r reviewed (approved):
LGTM 👍

mcdonnnj reviewed (changes requested):
Please refresh the information in the README to reflect your latest changes to the Terraform configuration.

jsf9k committed:
Update README.md to make it agree with these Terraform changes

mcdonnnj commented on variables.tf:
```diff
@@ -56,6 +56,30 @@ variable "aws_region" {
+
+variable "cloudwatch_policy_instances" {
+  type        = list(string)
+  description = "Each string corresponds to the name of an instance (as it appears in the CloudWatch log group) for which CDM is to be allowed read access to the specific CloudWatch log groups in which they are interested."
```
With the shift they're gaining access to specific streams now, right?

jsf9k replied:
Please have a look at the changes in e1df8a3 and let me know if they make things clearer.  They are gaining access to specific streams, but those streams reside inside log groups whose names are determined by the instance names.

mcdonnnj replied:
Oh yeah I understood what was happening, just the tail
> allowed read access to the specific CloudWatch log groups in which they are interested.

didn't seem to line up with how things worked now.

jsf9k committed:
Elaborate two variable descriptions to make them clearer

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>
